src/lando/main/scm/git.py

```diff
@@ -0,0 +1,298 @@
+                result.stdout,
+                result.stderr,
+            )
+        return result
+
+    def apply_patch(
+        self, diff: str, commit_description: str, commit_author: str, commit_date: str
+    ):
+        """Apply the given patch to the current repository
+
+        Args:
+            patch_buf (StringIO): The patch to apply
+
+        Returns:
+            None
+        """
+        self._git_apply_patch(diff, commit_description, commit_author, commit_date)
+
+    @contextmanager
+    def for_pull(self) -> ContextManager:
+        """Context manager to prepare the repo with the correct environment variables set for pulling."""
+        yield self
+
+    @contextmanager
+    def for_push(self, requester_email: str) -> ContextManager:
+        """Context manager to prepare the repo with the correct environment variables set for pushing.
+
+        Args:
+            requester_email (str)
+        """
+        # We set the committer name to the requester's _email_ as this is the only piece of information about the user that we are comfortable making public. Names in the User objects are coming from LDAP, and may not be accepblate to use publicly.
+        os.environ[ENV_COMMITTER_NAME] = requester_email
+        os.environ[ENV_COMMITTER_EMAIL] = requester_email
+        logger.debug(
+            f"{ENV_COMMITTER_EMAIL} and {ENV_COMMITTER_NAME} set to {requester_email}"
+        )
+        try:
+            yield self
+        finally:
+            del os.environ[ENV_COMMITTER_NAME]
+            del os.environ[ENV_COMMITTER_EMAIL]
+
+    def head_ref(self) -> str:
+        """Get the current revision_id"""
+        return self._git_last_commit_id()
+
+    def changeset_descriptions(self) -> list[str]:
+        """Retrieve the descriptions of commits in the repository.
+
+        Returns:
+            list[str]: A list of first lines of changeset descriptions.
+        """
+        command = ["log", "--format=%s", "@{u}.."]
+        return self._git_run(*command).splitlines()
+
+    def update_repo(self, pull_path: str, target_cset: Optional[str] = None) -> str:
+        """Update the repository to the specified changeset.
+
+        This method uses the Git commands to update the repository
+        located at the given pull path to the specified target changeset.
+
+        Args:
+            pull_path (str): The path to pull from.
+            target_cset (str): The target changeset to update the repository to.
+
+        Returns:
+            str: The target changeset
+        """
+        branch = target_cset or self.default_branch
+        self.clean_repo()
+        self._git_run("pull", "--prune", pull_path)
+        self._git_run("checkout", "--force", "-B", branch)
+
+    def clean_repo(self, *, strip_non_public_commits=True):
+        """Reset the local repository to the origin"""
+        if strip_non_public_commits:
+            self._git_run("reset", "--hard", "origin/HEAD")
```

Comment:
Is using `origin/HEAD` correct here? My thought is we would probably want to use `origin/<branch>` for whichever branch we're working on (`origin/beta`, `origin/central` etc). Is there any guarantee that `origin/HEAD` is the correct commit to reset to?

As an alternative, did you consider creating a temp branch to work with locally while we apply patches? For example we create a `worker_D12345` branch and apply patches on it, then to rollback we just `git branch -D worker_D12345`, or `git branch autoland HEAD -f` to move the branch to the final destination on push.

Reply:
Yeah, `origin/HEAD` may not always be correct. It's probably safer to use `origin/{self.default_branch}`

I didn't consider creating temporary branches, but I think this will add overhead in garbage-collecting them, which may not be worth the effort.